File: src/main/java/io/saso/dash/services/Service.java
package io.saso.dash.services;

import io.netty.channel.ChannelHandlerContext;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

public abstract class Service implements IService
{
    private final List<Pollable> subPollables = new CopyOnWriteArrayList<>();

    @Override
    public void start(ChannelHandlerContext ctx, DBEntityProvider db)
            throws Exception
    {
        for (Pollable subPollable : subPollables) {
            subPollable.start(ctx, db);
        }

        ctx.channel().flush();
    }

    @Override
    public void poll(ChannelHandlerContext ctx, DBEntityProvider db)
            throws Exception
    {
        for (Pollable subPollable : subPollables) {
            subPollable.poll(ctx, db);
        }

        ctx.channel().flush();
    }

    @Override
    public void stop(ChannelHandlerContext ctx, DBEntityProvider db)
            throws Exception
    {
        for (Pollable subPollable : subPollables) {
            subPollable.stop(ctx, db);
        }

        ctx.channel().flush();
    }

    protected final void registerSubPollable(Pollable pollable)
    {
        subPollables.add(pollable);
    }

    protected final void unregisterSubPollable(Pollable pollable)
    {
        subPollables.remove(pollable);
    }
}

File: src/main/kotlin/io/saso/dash/services/Service.kt
package io.saso.dash.services

import io.netty.channel.ChannelHandlerContext
import io.saso.dash.database.DBEntityProvider
import java.util.concurrent.CopyOnWriteArrayList

public abstract class ServiceOLD : ServicePollable
{
    private val pollables = CopyOnWriteArrayList<ServicePollable>()

    override fun start(ctx: ChannelHandlerContext, db: DBEntityProvider)
    {
        pollables.forEach {
            it.start(ctx, db)
        }

        ctx.channel().flush()
    }

    override fun poll(ctx: ChannelHandlerContext, db: DBEntityProvider)
    {
        pollables.forEach {
            it.poll(ctx, db)
        }

        ctx.channel().flush()
    }

    override fun stop(ctx: ChannelHandlerContext)
    {
        pollables.forEach {
            it.stop(ctx)
        }

        ctx.channel().flush()
    }

    abstract val pollInterval: Int

    protected fun registerSubPollable(pollable: ServicePollable)
    {
        pollables.add(pollable)
    }

    protected fun unregisterSubPollable(pollable: ServicePollable)
    {
        pollables.remove(pollable)
    }
}
